content/CustomExceptions/CustomExceptions.q.md

```diff
@@ -0,0 +1,1501 @@
+
+Correct the code so that it outputs the string `Exception caught: A negative number as a divisor is encountered in /Main.php`.
+
+```php
+<?php
+require_once(__DIR__ . "/DivideByZeroException.php");
+
+function divide($dividend, $divisor)
+{
+    if ($divisor == 0) {
+        throw new DivideByZeroException("The value '0' as a divisor is not allowed.");
+    } elseif ($divisor < 0) {
+        throw new DivideByNegativeException();
+    } else {
+        return ($dividend / $divisor);
+    }
+}
+
+try {
+    echo divide(10, -2);
+} catch (DivideByZeroException $div) {
+    echo "Exception caught: " . $div->errorMessage();
+} catch (DivideByNegativeException $neg) {
+    echo "Exception caught: " . $neg->myErrorMessage() . " in " . $neg->myErrorFile();
+}
+?>
+```
+
+```php
+<?php
+class DivideByZeroException extends Exception
+{
+    public function errorMessage()
+    {
+        return $this->getMessage();
+    }
+}
+?>
+```
+
+```php
+<?php
+class DivideByNegativeException extends Exception 
+{
+    public function myErrorMessage()
+    {
+        return "A negative number as a divisor is encountered";
+    }
+
+    public function myErrorFile()
+    {
+        return $this->getFile();
+    }
+}
+?>
+```
+
+
+:::
+
+/// type=REPL, readonly=true, filename=[Main.php,DivideByZeroException.php,DivideByNegativeException.php]
+
+```php
+<?php
+require_once(__DIR__ . "/DivideByZeroException.php");
+require_once(__DIR__ . "/DivideByNegativeException.php");
+
+function divide($dividend, $divisor)
+{
+    if ($divisor == 0) {
+        throw new DivideByZeroException("The value '0' as a divisor is not allowed.");
+    } elseif ($divisor < 0) {
+        throw  DivideByNegativeException();
+    } else {
+        return ($dividend / $divisor);
+    }
+}
+
+try {
+    echo divide(10, -2);
+} catch (DivideByZeroException $div) {
+    echo "Exception caught: " . $div->errorMessage();
+} catch (DivideByNegativeException $neg) {
+    echo "Exception caught: " . $neg->myErrorMessage() . " in " . $neg->myErrorFile();
+}
+?>
+```
+
+```php
+<?php
+class DivideByZeroException extends Exception
+{
+    public function errorMessage()
+    {
+        return $this->getMessage();
+    }
+}
+?>
+```
+
+```php
+<?php
+class DivideByNegativeException extends Exception 
+{
+    public function myErrorMessage()
+    {
+        return "A negative number as a divisor is encountered";
+    }
+
+    public function myErrorFile()
+    {
+        return $this->getFile();
+    }
+}
+?>
+```
+/// type=MS, answer=[4,5]
+
+Execute the program. What are the error messages?
+
+ - Undefined variable: `new` in /Main.php on line number 7
+
+ - Error: Class `'DivideByNegativeException'` not found in /Main.php on line number 9
+
+ - Uncaught Error: Class `'DivideByNegativeException'` not found in /Main.php on line number 9
+
+ - Error: Call to undefined function `DivideByNegativeException()` in /Main.php on line number 10
```

Comment:
Similar with my comment on line 950.

![image](https://user-images.githubusercontent.com/57437877/76546456-76f08d80-64c6-11ea-9653-103b87918ac5.png)

Reply:
the same explanation to the above. Please check. Thank you.

![image](https://user-images.githubusercontent.com/25945790/76576306-b2a64a00-64fc-11ea-9adf-54bf2bc9dc9f.png)